### ISSUE TITLE
[dev] Request: parallel tasks

### ISSUE THREAD

**Issue by YvesHenri** [PokemonGoF/PokemonGo-Bot#2221]
Some task needs to be run asynchronously, for example (given the example config file):

> The following configuration tells the bot to transfer all the Pokemon that match the transfer configuration rules, then recycle the items that match its configuration, then catch the pokemon that it can, so on, so forth. Note the last two tasks, MoveToFort and FollowSpiral. When a task is still in progress, it won't run the next things in the list. So it will move towards the fort, on each step running through the list of tasks again. **Only when it arrives at the fort and there are no other stops available for it to move towards will it continue to the next step and follow the spiral**.

```
{
  // ...
  "tasks": [
    {
      "type": "TransferPokemon"
    }, 
    {
      "type": "RecycleItems"
    }, 
    {
      "type": "CatchVisiblePokemon"
    }, 
    {
      "type": "CatchLuredPokemon"
    }, 
    {
      "type": "SpinFort"
    }, 
    {
      "type": "MoveToFort"
    }, 
    {
      "type": "FollowSpiral"
    }
  ]
  // ...
}
```

The problem is with MoveToFort. Imagine youre in a place with some forts, as well as a high pokemon spawn rate. It will not catch pokemons UNTILL all the forts have been visited, which is a problem.
Please post your thoughts about this below.

**Comment by Nik0lashka** [comment]
@YvesHenri check #2217

**Comment by Anakin5** [comment]
Move to fort will only do one step toward a fort. All tasks before MoveToFort are going to be executed between each step.

**Comment by xorinzor** [comment]
duplicate of #2054

**Comment by YvesHenri** [comment]
@xorinzor not really. He wants to run EVERYTHING asynchronously, like spin + catch + recycle all at once. This is not what I'm looking for. #2217 it what it looks like.

**Comment by xorinzor** [comment]
running spin catch and recycle at once should have no impact, so you might as well just do those all at once.

**Comment by YvesHenri** [comment]
@xorinzor you cant spin + catch + recycle at once in real life. It is impossible.

**Comment by xorinzor** [comment]
@YvesHenri maybe not all at once, but you can divide those into it's own worker-thread without them having an impact on the others.

I just don't see the point why you'd only run a couple in parallel if you can multithread a bunch more.

**Comment by YvesHenri** [comment]
> running spin catch and recycle at once should have no impact
You are contradicting yourself:
> maybe not all at once

Obviously, all these tasks could be parallelized. This, however, would not imply a humanly-behavior, leading to a possible ban. Bots are meant to be artificial humans (AI), not a superb killing machine giant monster.

Please check #2217.

**Comment by xorinzor** [comment]
I wasn't contradicting myself, I was correcting myself, agreeing with you that all at once wouldn't be the smartest thing.

However; I still think those tasks could run perfectly in their own worker-thread.

I don't see how else you would divide the tasks into their own threads without having the same issue.

Also, I don't understand what the pull request you're referring to has anything to do with this.

**Comment by YvesHenri** [comment]
You first claimed it was a duplicate of #2054. It is not.
The #2217 pull request basically solves this issue, as we can see below.

> When it is working it stops the move_to_fort worker because the pokemons have a limited lifetime and the bot cannot run the other worker simultaneously.

Also, I am not saying that they would impact on other multithreaded tasks or not. I'm saying that multithreading them all, just for the fact that its programatically possible, is not the smartest thing. If it were, we wouldnt need any randomized factors (catch_randomize_reticle_factor/catch_randomize_spin_factor) at all, for example.

Again, what I'm asking for is to parallelize the MoveToFort task, mainly. This is the longer task in the task chain and it doesn't allow to capture stuff untill its not finished (even if there are stuff around).

**Comment by xorinzor** [comment]
It's not multithreading them all, it's dividing them into 2 threads, that too is multithreading.

I understand what you're asking for, and I'm telling you it's a duplicate of #2054 since that guy wants the same thing.
